cordapp-example/workflows-kotlin/src/main/kotlin/com/example/flow/PaybackFlow.kt

```diff
@@ -0,0 +1,153 @@
+                    FINALISING_TRANSACTION
+            )
+        }
+
+        override val progressTracker = tracker()
+
+        /**
+         * The flow logic is encapsulated within the call() method.
+         */
+        @Suspendable
+        override fun call(): SignedTransaction {
+            progressTracker.currentStep = CREATING_COMMAND
+            //val signedTransaction = serviceHub.validatedTransactions.getTransaction(loanTxId)
+            // val signedTransaction = transactions.find { it.id == transactionHash } ?: throw IllegalArgumentException("Unknown transaction hash.")
+
+            /* val criteria = QueryCriteria.VaultQueryCriteria(
+                    contractStateTypes = setOf(IOUState::class.java),
+                    status = Vault.StateStatus.UNCONSUMED,
+                    exactParticipants = listOf(ourIdentity, otherParty)
+            ) // */
+
+            val uuid: UUID = UUID.fromString(inputLinearId)
+            val queryCriteria: QueryCriteria = LinearStateQueryCriteria().withUuid(Collections.singletonList(uuid))
+
+            val results = serviceHub.vaultService.queryBy(IOUState::class.java, queryCriteria)
+            // val ourStateRef = StateRef(SecureHash.parse(loanTxId), 0)
+            // val (state, ref) = serviceHub.toStateAndRef<ContractState>(ourStateRef)
+            val inputStateAndRef = serviceHub.toStateAndRef<ContractState>(results.states[0].ref)
+            val inputIOUState = inputStateAndRef.state.data as IOUState
+
+            progressTracker.currentStep = GENERATING_TRANSACTION
+            // Generate an unsigned transaction.
+            val txCommand = Command(IOUContract.Commands.Destroy(), inputIOUState.participants.map { it.owningKey })
+
+            // Obtain a reference to the notary that was in use for input state
+            val notary = results.states[0].state.notary
+
+            // build tx that will consume previous output
+            val txBuilder = TransactionBuilder(notary)
+                    .addInputState(inputStateAndRef)
+                    .addCommand(txCommand)
+
+            progressTracker.currentStep = VERIFYING_TRANSACTION
+            // Verify that the transaction is valid.
+            txBuilder.verify(serviceHub)
+
+            progressTracker.currentStep = SIGNING_TRANSACTION
+            // Sign the transaction (must be by a borrower)
+            val signedTx = serviceHub.signInitialTransaction(txBuilder)
+
+            progressTracker.currentStep = INITIALIZE_OTHER_PARTY_FLOW
+            // Send the state to the counterparty, and receive it back with their signature.
+            val lenderPartySession = initiateFlow(inputIOUState.lender)
+
+            val confirmPayback = lenderPartySession
+                    .sendAndReceive<Boolean>("ready for receiving payback?")
+                    .unwrap { it }
+
+            require(confirmPayback) { "Oops, lender failed to be ready" }
+
+            lenderPartySession.send(paybackValue) // */
```

Comment:
@adelRestom I wanted to use some send/receive communication as you suggested, and I wanted to use it in two different ways: `sendAndReceive` and `send`. Without this communication input state is consumed successfully, but with it I'm getting flow errors. I'm not able to figure out - could you please help me with this?

Reply:
Using `send`, `receive`, `sendAndReceive` requires a very strict orchestration between the initiator and responder, meaning for every `send` from the initiator; there should be a `receive` in the responder, and vice versa; otherwise one of them will be stuck on a `receive` that doesn't have a matching `send` from the counter party. So all `send` / `receive` should appear in the proper order in initiator / responder, for instance one mistake I see is:  
- Your initiator has `sendAndReceive<Boolean>("ready for receiving payback?")` meaning your initiator is doing a `send("ready for receiving payback?")`.  
- But your responder doesn't have a matching `receive` to receive that String, instead it immediately starts with a `send(true)`, so the initiator is stuck because it has a `send` that doesn't have a matching `receive` in the responder.  

So that the idea, for every `send` there should be ` receive` and they should appear in the right order.